lib/Sema/DerivedConformanceCaseIterable.cpp

```diff
@@ -0,0 +1,164 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file implements implicit derivation of the CaseIterable protocol.
+//
+//===----------------------------------------------------------------------===//
+
+#include "TypeChecker.h"
+#include "swift/AST/Decl.h"
+#include "swift/AST/Stmt.h"
+#include "swift/AST/Expr.h"
+#include "swift/AST/Types.h"
+#include "llvm/Support/raw_ostream.h"
+#include "DerivedConformances.h"
+
+using namespace swift;
+using namespace DerivedConformance;
+
+/// Common preconditions for CaseIterable.
+static bool canDeriveConformance(NominalTypeDecl *type) {
+  // The type must be an enum.
+  auto enumDecl = dyn_cast<EnumDecl>(type);
+  if (!enumDecl)
+    return false;
+
+  // The enum must not have associated values.
+  return !enumDecl->hasPotentiallyUnavailableCaseValue()
+      && enumDecl->hasOnlyCasesWithoutAssociatedValues();
+}
+
+/// Derive the implementation of allCases for a "simple" no-payload enum.
+void deriveCaseIterable_enum_getter(AbstractFunctionDecl *funcDecl) {
+  auto *parentDC = funcDecl->getDeclContext();
+  auto *parentEnum = parentDC->getAsEnumOrEnumExtensionContext();
+  auto enumTy = parentEnum->getDeclaredType();
+  auto &C = parentDC->getASTContext();
+
+  SmallVector<Expr *, 8> ElExprs;
+  for (EnumElementDecl *elt : parentEnum->getAllElements()) {
+    auto *ref = new (C) DeclRefExpr(elt, DeclNameLoc(), /*implicit*/true);
+    auto *base = TypeExpr::createImplicit(enumTy, C);
+    auto *apply = new (C) DotSyntaxCallExpr(ref, SourceLoc(), base);
+    ElExprs.push_back(apply);
+  }
+  auto *arrayExpr = ArrayExpr::create(C, SourceLoc(), ElExprs, {}, SourceLoc());
+
+  auto *returnStmt = new (C) ReturnStmt(SourceLoc(), arrayExpr);
+  auto *body = BraceStmt::create(C, SourceLoc(), ASTNode(returnStmt),
+                                 SourceLoc());
+  funcDecl->setBody(body);
+}
+
+static BoundGenericType *computeAllCasesType(NominalTypeDecl *enumType) {
+  auto *arrayDecl = enumType->getASTContext().getArrayDecl();
+  assert(arrayDecl);
+  auto metaTy = enumType->getInterfaceType();
+  if (!metaTy || metaTy->hasError())
+    return nullptr;
+
+  return BoundGenericType::get(arrayDecl, Type(),
+                               {metaTy->getRValueInstanceType()});
+}
+
+static Type deriveRawRepresentable_AllCases(TypeChecker &tc, Decl *parentDecl,
```

Comment:
... you mean `deriveCaseIterable_AllCases`, not `deriveRawRepresentable_AllCases`